### ISSUE TITLE
Implement threading for Robot Framework core

### ISSUE THREAD

**Issue by test-fullautomation** [robotframework/robotframework#5294]
start of the discussion is here https://github.com/robotframework/robotframework/issues/5026#issuecomment-2541196289

Goal is a discussion and conclusion how threading for Robot Framework could look like.

Here the content from above link:

@pekkaklarck :
maybe if you have a little time you can look into this pull-request from  RobotFramework AIO: https://github.com/test-fullautomation/robotframework/pull/29/files
This is our idea how to realize threading in Robot Framework core. There are many pull-requests related to threading following for stabilization, but it shows the base idea.
Would be really nice to get a feedback or idea about this from you.
It's not yet stable enough that we would make a proposal based on 7.1, but If you would be interested we could increase our efforts. Of course we would also consider your findings/proposals.
I have resources where I can assign tasks.

amendment:
Basic idea is to have an as easy as possible way to fork a thread by means of a builtin keyword. The thread will be managed by the core.
In addition a builtin notification keyword is required which allows to exchange payload between threads. Finally a builtin keyword is required to wait for a certain notification.
When implementing our own test framework, I found that these features fulfilled all testing requirements we had in the last 15 years.
Libraries must be thread safe, too. We have a base class which implements generic queued connections. Each message received by a connection is received by a listener thread which deploys the messages to all threads. This allows parallel non-blocking waiting for messages or message-pattern. [robotframework-qconnect-base](https://github.com/test-fullautomation/robotframework-qconnect-base) is a port (from us) of this ideas for Robot Framework.

What are your feelings about this ideas?

**Comment by pekkaklarck** [comment]
First of all, #5026 is about running Robot itself in a thread, not about running keywords in threads. It would be great if both would be supported, but the former doesn't have too high priority as processes in general work better in such usage anyway.

Being able to run keywords in threads would be extremely valuable, but unfortunately getting there is a rather big task. My preferred approach would be allowing keywords to be run in parallel using a syntax like
```

**Comment by test-fullautomation** [comment]
Yes it's a huge task, and it needs time to trust the implementation. It needs time to stabilize this feature.

What is working for us and what we already implemented is the following syntax. 
You can see this and much more in our atest: https://github.com/test-fullautomation/robotframework/blob/develop_6.1/atest/robot/running/thread/thread.robot

THREAD can be of course also be called PARALLEL instead:
```
Thread executed
    THREAD    TEST_THREAD    False
        SLEEP  2
        LOG   inside threading
    END
    LOG    CONTINUE IN MAIN THREAD
    SLEEP    5
````

Communication between threads with any payload is realised with a thread safe queue per thread. Also specific threads can be addressed for notification due by means of thread names assigned when forking the thread:
```
    THREAD    TEST_THREAD    False
        SLEEP  2
        send thread notification        thread notify      Thread Done
    END
    ${var}=     wait_thread_notification       thread notify      5
    LOG      Get notification from thread. Payloads: ${var}
    Should contain    ${var}    Thread Done
```

w.r.t your points:
1. currently we write in a file per thread. Your proposal would be also possible. Challenge is to allow to filter log entries by thread. If the log is highly mixed due to multiple threads that's very important.
2. In case of threads the notification mechanism must be used. Direct writing variable access should be avoided. Notification payloads are thread safe.
3. Please refer to my example above.

**Comment by franzhaas** [comment]
## This is discussion input, not a suggestion to modify robotframework proper.

I want to pitch additional approaches to solve the problem of using robotframework from multiple concurrent activities, which are all cooperative in the robotframework domain, and some even on python/C level.

### Threads
I did put together a quick and ugly demonstrator on how to have full access to all robotframework APIs in a thread safe way from a python implemented keyword running in its own thread. There are keyword libraries available for this, but this code demonstrates the idea, and does not provide a solution.

https://github.com/franzhaas/parallel_execution_rf

While less powerful, it is also much less invasive, provides a very small attack surface for synchronization bugs, and solves the problems that many libraries, most notably tkinker (dialogs), have with threads.

### Processes
Using multiple robotframework processes and communicating between those using message queues (I did use zmq professionally, but I don't have a minimal example ready at the moment)

### Async approach/callbacks
Also no minimal example ready at the moment, but the pseudo-code would look like this 

```python

def async_callbacks_based_parallelism():
    # Starting of the threads generating the data  is not shown
    while notDonne:
        nameOfKeywordToCallWithResult, args, kwargs = thread_generated_nextSteps_queue.get()
        BuiltIn().run_keyword(nameOfKeywordToCallWithResult, *args, **kwargs)
```

This solution lends well to the use case.:
Wait for data from multiple sources (_NOT_ all of them are io, if all are io, consider select)
Then do something _quickly_
Then wait again

### Async approach / select
When waiting for IO from multiple sources, python provides a [select](https://docs.python.org/3/library/select.html) functionality which waits for multiple io's at once. An example is [here](https://github.com/MarketSquare/robotframework-construct/tree/main/atests/nfc_nci) although this particular example only uses one IO source.

## Conclusion
I believe that more discussion and documentation on this problem/solution space would be great. I hope such discussions may take place at robocon, and will do my part at the very least at the relevant workshop on this topic.

**Comment by franzhaas** [comment]
@test-fullautomation 

Yes windows has reduced select capability, but by chance the robotframework-construct library comes with an example for exactly the serial use case. No threads are needed.

```python
    def wait_for_data_from_nfc(self, timeout: float = 1.0):
        """
        Waits for data from the NFC device.

        Raises an exception if the NFC connection is not open.
        """
        if self._serial_connection and self._serial_connection.is_open:
            try:
                select.select([self._serial_connection.fileno()], [], [], timeout)
            except io.UnsupportedOperation:
                    # Windows does not support select on serial ports, so we have to do it the hard way
                    endTime = time.time() + timeout
                    while time.time() < endTime and not self._serial_connection.in_waiting:
                        time.sleep(0.001)
            assert self._serial_connection.in_waiting, "Timeout while waiting for data from NFC device"
        else:
            raise Exception("NFC connection is not open")
```
### About needing python skills
I guess you looked at [this](https://github.com/franzhaas/parallel_execution_rf/blob/main/parallel_execution_keyword.py)?

Technically Python source modifications are only needed to avoid changes to the robotframework core, which is a choice.

```python

def _start_guess_number_1_to_100(nr, _loginfoQueue):
    """
    This is a toy example of a long running task which can be executed in parallel.
    """
    trials = 0
    try:
        _loginfoQueue.put((logger.console, f"Start guess number {nr}",))
       #

```
is equivalent to


```python

def _start_guess_number_1_to_100(nr):
    """
    This is a toy example of a long running task which can be executed in parallel.
    """
    trials = 0
    try:
        logger.console(f"Start guess number {nr}")
---8<---8<---8<---8<---8<---
```

If your logger.console code looks like this
```python

def console(*args, **kwargs):
        _namespaceForConcurrencyQueues._loginfoQueue.put(("console", args, kwargs))

```
If the code on the other end of _loginfoQueue does log correctly.

**Comment by franzhaas** [comment]
If you are interested in this discussion, you might be interested in the developments around this issue.: 

https://github.com/test-fullautomation/robotframework/issues/110

**Comment by pekkaklarck** [comment]
Quick comments:
1. Parallel execution of keywords is an important topic.
2. I personally don't think Robot language should have thread primitives. I believe they are too complicated on this level.
3. The first step to get any syntax support for parallel execution is proposing a design how the syntax should work. Changing syntax afterwards is hard so we need to design it carefully. Implementation can be done later and it also can be changed without problems (as long as the old syntax works).
4. At the moment the best approach for parallel execution of keywords is doing that on a library level.

**Comment by test-fullautomation** [comment]
@pekkaklarck , @franzhaas ,
thank you for all your input. I will be in Christmas holiday for two weeks. After this I will directly process your questions and proposals.

Did you have time to look into: [threading atest](https://github.com/test-fullautomation/robotframework/blob/develop_6.1/atest/robot/running/thread/thread.robot)?
In our [reference documentation chapter 7 page 13](https://robotframework-aio.org/download/RobotFrameworkAIO_Reference_extended.pdf) you find also some information about syntax and BuiltIns.
To have an idea about the implementation you find the initial Robot Framework core changes [here](https://github.com/test-fullautomation/robotframework/pull/29/files) and related to BuiltIns [here](https://github.com/test-fullautomation/robotframework/pull/32). (There are mutliple pull-requests following for stabilization).
This is our proposal for syntax as well as BuiltIns related to threading. 
It shows how threads are forked, notifications can be sent and how to wait for notifications. Using our Robot Framework fork you can experiment with it to get a feeling how it works. 

We are looking forward you to your advice and ideas!

**I wish you a Happy Christmas and a Happy New Year!**

**Comment by pekkaklarck** [comment]
It's a lot of work to go through your changes to see what you actually have done. I would hope that you'd write a design document/proposal where you explain how you'd like this to work. The doc should be something similar to PEPs that are used with Python, but it doesn't need to be that format as we don't (yet) have a formal process for this kind of enhancements.

The above said, I did take a quick look at your changes. The added syntax looks rather complicated.

**Comment by test-fullautomation** [comment]
I got your point. We will workout a design document with explanations and background information beginning of next year.
Nice time!
